src/vs/workbench/api/browser/mainThreadLanguageFeatures.ts

```diff
@@ -251,9 +251,16 @@ export class MainThreadLanguageFeatures extends Disposable implements MainThread
 	// --- extra info
 
 	$registerHoverProvider(handle: number, selector: IDocumentFilterDto[]): void {
+		/*
+		const hoverFinalizationRegistry = new FinalizationRegistry((hoverId: number) => {
+			this._proxy.$releaseHover(handle, hoverId);
+		});
+		*/
 		this._registrations.set(handle, this._languageFeaturesService.hoverProvider.register(selector, <languages.HoverProvider>{
-			provideHover: (model: ITextModel, position: EditorPosition, token: CancellationToken): Promise<languages.Hover | undefined> => {
-				return this._proxy.$provideHover(handle, model.uri, position, token);
+			provideHover: async (model: ITextModel, position: EditorPosition, token: CancellationToken, context?: languages.HoverContext<{ id: number }>): Promise<HoverWithId | undefined> => {
+				const hover = await this._proxy.$provideHover(handle, model.uri, position, context, token);
+				// hoverFinalizationRegistry.register(hover, hover.id);
+				return hover;
 			}
```

Comment:
You want to implement `languages.HoverProvider<HoverWithId>` (and then use `languages.HoverContext<HoverWithId>` - what you currently do breaks types, as the default HoverProvider uses Hover, which does not have an id). Avoid casts here and use `... satisfies <TYPE>`!

Also, when you call `$provideHover`, you don't want pass in context.previousHover as is, but just send its id (to avoid serialization issues).
I think $provideHover takes `HoverContext<{ id: number }>`, not `HoverContext<HoverWithId>`. However, the `context` you get in this implementation is not just of the shape `{ id: number }`, but of the shape `HoverWithId`, since that is what you return when a normal hover is requested (and you get the last reference you returned).

Reply:
Since this is just a typing error and it compiles, it is not a blocker, but please fix this later.

Reply:
Yes I made the PR https://github.com/microsoft/vscode/pull/210889